Reenable cache

jpsim opened the pull request:
Continued from #1311. Addresses #1184.

jpsim commented:
I'd love if you could review this too @victorpimentel.

victorpimentel commented on Source/SwiftLintFramework/Models/LinterCache.swift:
```diff
@@ -18,58 +18,67 @@ public enum LinterCacheError: Error {
+
         lock.lock()
-        var filesCache = (cache["files"] as? [String: Any]) ?? [:]
+        var filesCache = (cache[LinterCache.Key.files.rawValue] as? [String: Any]) ?? [:]
```
Very small code style question: is there any reason to use LinterCache.Key in some lines and just Key in others? It's not important, but I'm reviewing this from my phone so I'm not 100% sure if they refer to another type.

jpsim replied:
I (inconsistently) omitted `LinterCache` within its body since it could be inferred. This one slipped through. Fixed in 43e4ed6.

marcelofabri commented on CHANGELOG.md:
```diff
@@ -11,8 +11,27 @@
 
+* Cache linter results for files unmodified since the previous linter run.
+  [Victor Pimentel](https://github.com/victorpimentel)
+  [#1184](https://github.com/realm/SwiftLint/issues/1184)
```
missing two trailing spaces on all entries 😅

jpsim replied:
new computer accidentally has `whitespace = fix` in its `.gitconfig` so it automatically trimmed trailing whitespace when rebasing. Promptly removed that.

marcelofabri commented on Source/SwiftLintFramework/Models/Configuration.swift:
```diff
@@ -34,9 +32,25 @@ public struct Configuration: Equatable {
+    public var cacheDescription: String {
+        let cacheRulesDescriptions: [String: Any] = rules.reduce([:]) { accu, element in
+            var accu = accu
+            accu[type(of: element).description.identifier] = element.configurationDescription
```
the `configurationDescription` of `ConditionalReturnsOnNewline` is `"N/A"` for some reason. We probably should change that as well.

jpsim replied:
fixed in #1523.

marcelofabri commented on Source/SwiftLintFramework/Models/Configuration.swift:
```diff
@@ -34,9 +32,25 @@ public struct Configuration: Equatable {
+            "root": rootPath ?? FileManager.default.currentDirectoryPath,
+            "rules": cacheRulesDescriptions
+        ]
+        if let prettyJSONData = try? JSONSerialization.data(withJSONObject: dict),
```
shouldn't this pass `options` to actually be a pretty? 😆

jpsim replied:
doesn't need to be pretty actually, since I'm just using this for serialization. fixed variable name.

marcelofabri commented:
Also, does this handle nested configurations? (It doesn't look like it does)

jpsim commented:
I've updated/rebased this after #1522, #1523 & #1524 were merged. I also updated it to consider custom rules in cache invalidation.

In manual testing, I've confirmed that:

An individual file cache is invalidated when:
- The file is modified, either by changing its contents or running `touch file.swift`

Entire cache is invalidated when:
- `.swiftlint.yml` custom rules are added/removed/modified
- `disabled_rules`, `enabled_rules`, `opt_in_rules`, `whitelist_rules` are added/removed/changed
- a rule's configuration is changed in `.swiftlint.yml`

Cache is not invalidated when:
- contents of `.swiftlint.yml` is reordered
- comments or whitespace are added/removed in `.swiftlint.yml`
- `cache_path`, `excluded`, `included`, `reporter`, `swiftlint_version`, `use_nested_configs` or `warning_threshold` are added/removed/changed

The only thing I can see doesn't work is nested configurations. I haven't really looked into what it would take to support that either. I suspect it wouldn't be too hard though.

marcelofabri commented on Source/SwiftLintFramework/Rules/RuleConfigurations/RegexConfiguration.swift:
```diff
@@ -27,6 +27,29 @@ public struct RegexConfiguration: RuleConfiguration, Equatable {
+        dict["message"] = message
+        dict["regex"] = regex.pattern
+        if let included = included?.pattern {
+            dict["included"] = included
```
don't `dict["included"] = included?.pattern` have the same effect?

jpsim replied:
that would produce `["included": nil]` rather than `[:]`

marcelofabri replied:
It doesn't 😅

```
var dict = [String: String]()
dict["foo"] = nil
dict["bar"] = "bar"

let data = try! JSONSerialization.data(withJSONObject: dict)
print(String(data: data, encoding: .utf8)!) // {"bar":"bar"}
```

jpsim replied:
👍  thanks

marcelofabri commented:
> The only thing I can see doesn't work is nested configurations. I haven't really looked into what it would take to support that either. I suspect it wouldn't be too hard though.

Do you want to merge this without supporting nested configurations? If so, I'd suggest to disable caching by default until we do support them.

jpsim commented:
WIP support for nested configurations: https://github.com/realm/SwiftLint/compare/jp-reenable-cache...jp-wip-cache-support-nested-configs

I have some cleaning up to do and tests to write, but it shows that it can be done fairly easily.

jpsim commented:
I'd really rather not ship this (again) with known issues or limitations.

samskiter commented:
Thanks so much for this @jpsim  👍

jpsim commented:
Superseded by #1530.
